File: 4_Food_Service/test_integration_food.js
// test_integration_food.js
const axios = require('axios');

const FOOD_SERVICE_URL = "http://localhost:8004";
const GAME_ID = "partie_test";
const DISPLAY_URL = "http://localhost:8010/display/on-food-eaten";

let receivedEvents = [];

// Simuler le Display Service en Express
const express = require('express');
const displayApp = express();
displayApp.use(express.json());

displayApp.post('/display/on-food-eaten', (req, res) => {
    console.log("[Display] Événement reçu:", req.body);
    receivedEvents.push(req.body);
    res.json({ status: "ok" });
});

const displayServer = displayApp.listen(8010, () => {
    console.log("Display Service en cours d'exécution sur le port 8010");
});

async function runTest() {
    try {
        // 1️⃣ Subscribe le Display Service au Food Service
        let r = await axios.post(`${FOOD_SERVICE_URL}/food/${GAME_ID}/subscribe`, { route: DISPLAY_URL });
        console.log("[Test] Subscribe:", r.data);

        // 2️⃣ Simuler un déplacement du serpent
        r = await axios.post(`${FOOD_SERVICE_URL}/food/${GAME_ID}/on-move`, { x: 5, y: 5 });
        console.log("[Test] On-Move:", r.data);

        // 3️⃣ Manger la nourriture
        r = await axios.delete(`${FOOD_SERVICE_URL}/food/${GAME_ID}`);
        console.log("[Test] Eat Food:", r.data);

        // 4️⃣ Attendre la notification
        await new Promise(resolve => setTimeout(resolve, 1000));
        console.log("[Test] Événements reçus par Display:", receivedEvents);
    } catch (err) {
        console.error(err);
    } finally {
        displayServer.close(() => console.log("[Test] Display Service arrêté."));
    }
}

runTest();

File: 4_Food_Service/test_integration_food.py
import asyncio
import httpx

FOOD_SERVICE_URL = "http://localhost:8004"
GAME_ID = "partie_test"

# --- Simule un Display Service ---
from fastapi import FastAPI
from pydantic import BaseModel
import uvicorn

display_app = FastAPI(title="Fake Display Service")

class FoodEvent(BaseModel):
    eventName: str
    gameId: str
    food: dict

received_events = []

@display_app.post("/display/on-food-eaten")
async def on_food_eaten(event: FoodEvent):
    print(f"[Display] Événement reçu: {event}")
    received_events.append(event)
    return {"status": "ok"}

async def run_display_service():
    config = uvicorn.Config(display_app, host="0.0.0.0", port=8010, log_level="info")
    server = uvicorn.Server(config)
    await server.serve()

# --- Tests d'intégration ---
async def test_food_service():
    async with httpx.AsyncClient() as client:
        display_url = "http://localhost:8010/display/on-food-eaten"
        
        # Subscribe le Display Service au Food Service
        r = await client.post(f"{FOOD_SERVICE_URL}/food/{GAME_ID}/subscribe", json={"route": display_url})
        print("[Test] Subscribe:", r.json())
        
        # Simuler un déplacement du serpent
        r = await client.post(f"{FOOD_SERVICE_URL}/food/{GAME_ID}/on-move", json={"x": 5, "y": 5})
        print("[Test] On-Move:", r.json())
        
        # Manger la nourriture
        r = await client.delete(f"{FOOD_SERVICE_URL}/food/{GAME_ID}")
        print("[Test] Eat Food:", r.json())

        # Vérifier que l'événement est bien reçu
        await asyncio.sleep(1)  # attendre la notification
        print(f"[Test] Événements reçus par Display: {received_events}")

# --- Main ---
async def main():
    # Lancer le fake Display Service en tâche séparée
    display_task = asyncio.create_task(run_display_service())
    
    # Attendre 1 seconde pour que le serveur démarre
    await asyncio.sleep(1)
    
    # Lancer le test Food Service
    await test_food_service()
    
    # Arrêter le serveur après le test
    display_task.cancel()
    try:
        await display_task
    except asyncio.CancelledError:
        print("[Test] Display Service arrêté.")

if __name__ == "__main__":
    asyncio.run(main())
